Sprint 13

SelickSD opened the pull request.

SelickSD committed:
Вынес во WebViewPresenter функционал из WebViewViewController по инструкции из лекции

SelickSD committed:
Сделал тесты WebViewTests

SelickSD committed:
Вылечил острые углы

SelickSD committed:
Изменил структуру проекта

SelickSD committed:
Делаю рефакторинг
Перенес ответственность по отрисовки View в ProfileView

SelickSD committed:
Сделал рефакторинг для ProfileViewController

SelickSD committed:
Добавил profileImageServiceObserver

SelickSD committed:
Добавил removeObserver

SelickSD committed:
Сделал рефакторинг ImagesListViewController

SelickSD committed:
Сделал Юнит тесты

SelickSD committed:
Завершил задание

DanilaG commented on YaImageFeed/ViewControllers/AuthAndWebView/WebViewViewController.swift:
```diff
@@ -0,0 +1,78 @@
+        presenter?.viewDidLoad()
+    }
+
+    override func observeValue(forKeyPath keyPath: String?, of object: Any?, change: [NSKeyValueChangeKey : Any]?, context: UnsafeMutableRawPointer?) {
```
Давай разобьём строчку на несколько. Это поможет сделать код более читаемым.

Так, например, гугл [в своём кодстайле к swift](https://google.github.io/swift/#:~:text=Swift%20code%20has%20a%20column%20limit%20of%20100%20characters.) рекомендует использовать строки не более 100 символов в длину

SelickSD replied:
Добрый вечер

По поводу замечания что у первой картинки в ленте отсутсвуют скругления, давай добавим их.

Я с этой проблемой уже пробовал бороться, задавал скругление и в сториборд и при создании ячейки и после создания ячейки и в кингфишере - ничего не помогает. Я предполагаю что спасет только удаление сториборда и верстка кодом.. 

Задавал вопрос в чате в Пачке, там предложили удалить все скругления и оставить только в awakeFromNib.. я так и сделал, но проблема не ушла.. хотя она появляется не каждый раз.. 


По поводу замечания: При скроле ленты присутсвуют фризы, попробуй их убрать - тоже пробовал скачивать картинки уменьшенного размера, но это не помогло, не знаю как быть, в пачке тоже посоветовали мелкие картинки качать, но это не решает проблемы.

DanilaG replied:
Привет! Попробовал часок поэксперементировать с проблемами, выглядит как магия - значения есть, а на экране не отображается)
Может быть действительно каке-то проблема в KF.

Тут могу посоветовать 2 подхода:
- начинать отключать определённую функциональность на данном экране и смотреть на результат, чтобы выявить, какой именно функционал даёт такое поведение
- начать с начала, создав вторую версию экрана и начинать добавлять постепенно функционал, до тех пор пока проблема не появится. Тут самое главное делать как можно проще. И потом постепенно усложнять до появления проблемы

DanilaG commented on YaImageFeed/ViewControllers/AuthAndWebView/WebViewViewController.swift:
```diff
@@ -0,0 +1,78 @@
+    }
+
+    @IBAction private func didTapBackButton(_ sender: Any) {
+        delegate?.webViewViewControllerDidCancel(self)
```
Я бы делегат для навигации также вынес в презентер + добавил бы в презентер метод на клик назад. Так как со временем, различные кнопки имеют способность обрастать логикой и прочими сущностями которые должны жить в презентере. Самый частый кейс - добавить аналитику на тап по кнопке)

DanilaG commented on YaImageFeed/Security/AuthConfiguration.swift:
```diff
@@ -0,0 +1,44 @@
+let AccessScope: String = "public+read_user+write_likes"
+
+let DefaultBaseURL = URL(string: "https://api.unsplash.com")!
+let UnsplashAuthorizeURLString = "https://unsplash.com/oauth/authorize"
```
Сейчас, к одним и тем же данным можно получить доступ 2 путями - через глобальные константы и `standard` структуру. Что со временем может вызывать путоници (особенно если проект будут поддерживать другие разработчики)

Давай избавимся от глобальных констант, сохранив значение напрямую в`standard`, а там где они нужны, будем использовать через `standard` структуру

DanilaG commented on YaImageFeed/Security/AuthConfiguration.swift:
```diff
@@ -0,0 +1,44 @@
+                                     accessScope: AccessScope,
+                                     authURLString: UnsplashAuthorizeURLString,
+                                     defaultBaseURL: DefaultBaseURL)
+        }
```
Давай сделаем получение стандартной структуры не через computed property, а через хранимую константу (`static let`). Так как при использовании computed property, если отсутсвуют оптимизации, структура будет создоваться каждый раз при обращении к `standard`. Но так как в ней ничего не меняется, то можно её сделать хранимой, и тогда она создастя только единожды

DanilaG commented on YaImageFeedTests/WebViewTests/WebViewTests.swift:
```diff
@@ -0,0 +1,98 @@
+    func testCodeFromURL() {
+        var urlComponents = URLComponents(string: "https://unsplash.com/oauth/authorize/native")!
+        urlComponents.queryItems = [
+            URLQueryItem(name: "code", value: "test code")
```
Давай вынесим `"test code"` в константу. Так как при изменении значения в этом месте, нам придётся менять значение на 94 строчке

DanilaG commented on YaImageFeedTests/ProfileViewTests/ProfileViewTests.swift:
```diff
@@ -0,0 +1,47 @@
+        view.presenter = presenter
+
+        
+
```
Давай оставим только 1 вертикальный отступ)

DanilaG commented on YaImageFeed/Protocols/ImagesListPresenterProtocol.swift:
```diff
@@ -0,0 +1,21 @@
+//  Created by Сергей Денисенко on 30.10.2023.
+//
+
+import UIKit
```
В данном случае презентер зависит от UI фреймворка, что не очень хорошо( Давай избавимся от этого

DanilaG commented on YaImageFeed/ViewControllers/Navigation/TabBarController.swift:
```diff
@@ -0,0 +1,33 @@
+
+        let profileViewController = ProfileViewController()
+        let profileViewPresenter = ProfileViewPresenter()
+        profileViewPresenter.view = profileViewController
+        profileViewController.presenter = profileViewPresenter
```
Retain cycle. Презентер сильная ссылка на view, view сильная ссылка на презентер

DanilaG commented on YaImageFeed/ViewControllers/Navigation/TabBarController.swift:
```diff
@@ -0,0 +1,33 @@
+
+        let imagesListViewController = storyboard.instantiateViewController(withIdentifier: "ImagesListViewController") as! ImagesListViewController
+        let imagesListPresenter = ImagesListPresenter()
+        imagesListPresenter.view = imagesListViewController
+        imagesListViewController.presenter = imagesListPresenter
```
Retain cycle. Презентер сильная ссылка на view, view сильная ссылка на презентер

DanilaG commented on YaImageFeed/ViewControllers/Navigation/TabBarController.swift:
```diff
@@ -0,0 +1,33 @@
+        super.awakeFromNib()
+        let storyboard = UIStoryboard(name: "Main", bundle: .main)
+
+        let imagesListViewController = storyboard.instantiateViewController(withIdentifier: "ImagesListViewController") as! ImagesListViewController
```
Давай избавимся от force cast, так как они могут приводить к падению приложения

SelickSD committed:
Убрал retain cycle  для ProfileViewController и ImagesListViewController
Поменял цвет под страницей профиля на YP Black
Делегат для навигации вынес в презентер

SelickSD committed:
Убрал глобальные переменные

SelickSD committed:
Убрал зависимость от фреймворка UIKit
Сделал статическую константу в  AuthConfiguration
Убрал force cast в TabBarController

SelickSD committed:
Сделал создание tableView раньше запуска viewDidLoad в presenter

SelickSD committed:
Исправил тесты

SelickSD committed:
Попробовал оптимизировать код и убрать фризы, но не особо получилось

SelickSD committed:
1
